tests/Builder/BuiderCodecTest.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+namespace MongoDB\Tests\Builder;
+
+use Generator;
+use MongoDB\Builder\Aggregation;
+use MongoDB\Builder\BuilderCodec;
+use MongoDB\Builder\Expression;
+use MongoDB\Builder\Pipeline;
+use MongoDB\Builder\Query;
+use MongoDB\Builder\Stage;
+use MongoDB\Tests\TestCase;
+
+use function array_merge;
+
+class BuiderCodecTest extends TestCase
+{
+    public function testPipeline(): void
+    {
+        $pipeline = new Pipeline(
+            Stage::match(Aggregation::eq('$foo', 1)),
+            Stage::match(Aggregation::ne('$foo', 2)),
+            Stage::limit(1),
+        );
+
+        $expected = [
+            (object) ['$match' => (object) ['$eq' => ['$foo', 1]]],
+            (object) ['$match' => (object) ['$ne' => ['$foo', 2]]],
+            (object) ['$limit' => 1],
+        ];
+
+        $this->assertSamePipeline($expected, $pipeline);
+    }
+
+    /** @see https://www.mongodb.com/docs/manual/reference/operator/aggregation/match/#perform-a-count */
+    public function testPerformCount(): void
+    {
+        $pipeline = new Pipeline(
+            Stage::match(Query::or(
+                ['score' => [Query::gt(70), Query::lt(90)]],
+                ['views' => Query::gte(1000)],
+            )),
+            Stage::group(null, ['count' => Aggregation::sum(1)]),
```

Comment:
Consider the named argument form:

```suggestion
            Stage::group(
                _id: null, 
                count: Aggregation::sum(1)
            ),
```

Reply:
There's also a typo in this class name (or maybe just the file name). Should be: BuilderCodecTest.

Reply:
The implementation you proposed for `Stage::group` factory.
```php
    public static function group(mixed $_id = null, array|null|object ...$fields): GroupStage
    {
        if (is_array($fields) && count($fields) === 1 && array_key_first($fields) === 0) {
            $fields = $fields[0];
        }

        return new GroupStage($_id, $fields);
    }
```

This will mess with types because we will have to accept `array|null|object` or a single `array<array|null|object>`.

We can say that `array` are not accepted and replace named arguments by `...[]` syntax when dotted field names are necessary.

```php
Stage::group(
    _id: null, 
    ...['count' => Aggregation::sum(1)]
),
```

Reply:
Playground: https://3v4l.org/QBBie

Reply:
Tracked in [PHPLIB-1267](https://jira.mongodb.org/browse/PHPLIB-1267)

Reply:
I'm trying to wrap my head around the conflict between named args and unpacking:

```
/* PHP 8.1+ allows this and merges the named argument into unpacked args.
 * PHP 8.0 yields a fatal error ("Cannot combine named arguments and argument unpacking") */
acceptsVariadicNamed(...['foo.bar' => 'bar', 'bar' => 'baz'], baz: 'bla');

/* PHP 8.0+ yields a fatal error ("Cannot use argument unpacking after named arguments") */
acceptsVariadicNamed(baz: 'bla', ...['foo.bar' => 'bar', 'bar' => 'baz']);
```

I recall you proposed requiring PHP 8.1 for the query builder (might have been via Slack), so let's assume the first call style above is used.

Is that still problematic for `$group` because it would require expressing all fields _before_ the required `$_id` parameter?

> This will mess with types because we will have to accept `array|null|object` or a single `array<array|null|object>`.

If each optional field must be an object of the form `{ <accumulator1> : <expression1> }`, I have two questions:

 * Why would `null` be accepted?
 * If an `array` is accepted, wouldn't the actual type be an array with a single string key corresponding to an [accumulator operator](https://www.mongodb.com/docs/manual/reference/operator/aggregation/group/#std-label-accumulators-group)? In that case, wouldn't that be suitable to distinguish the type from a list of `array|object` values (at least in Psalm)? 

https://www.mongodb.com/docs/manual/reference/operator/aggregation/group/#accumulator-operator

Reply:
Discussed over video. So there's no need to demonstrate the following in docs:

```
acceptsVariadicNamed(...['foo.bar' => 'bar', 'bar' => 'baz'], baz: 'bla');
```

Users can utilize named arguments or utilize array unpacking as they wish, and we don't have to encourage them to mix both forms.

Reply:
Also confirmed that the argument is now a map of expressions so my earlier question about accepting `null` no longer applies.